File: packages/python-sdk/strait/operations/rbac.py
"""Role-based access control operations."""

from __future__ import annotations

from typing import Any

from strait.operations._base import AsyncBaseService, BaseService


class RBACService(BaseService):
    def list_audit_events(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/audit-events", query=query)

    def list_members(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/members", query=query)

    def create_member(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/members", body=body)

    def delete_member(self, user_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/members/{userID}", path_params={"userID": user_id},
        )

    def bulk_member(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/members/bulk", body=body)

    def list_roles(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/roles", query=query)

    def create_role(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/roles", body=body)

    def get_role(self, role_id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/roles/{roleID}", path_params={"roleID": role_id},
        )

    def update_role(self, role_id: str, body: Any) -> dict[str, Any]:
        return self._request(
            "PATCH", "/v1/roles/{roleID}", path_params={"roleID": role_id}, body=body,
        )

    def delete_role(self, role_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/roles/{roleID}", path_params={"roleID": role_id},
        )

    def list_resource_policies(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return self._request("GET", "/v1/resource-policies", query=query)

    def create_resource_policy(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/resource-policies", body=body)

    def delete_resource_policy(self, policy_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/resource-policies/{policyID}",
            path_params={"policyID": policy_id},
        )

    def list_tag_policies(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/tag-policies", query=query)

    def create_tag_policy(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/tag-policies", body=body)

    def delete_tag_policy(self, policy_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/tag-policies/{policyID}", path_params={"policyID": policy_id},
        )

    def seed_roles(self) -> dict[str, Any]:
        return self._request("POST", "/v1/seed-roles")


class AsyncRBACService(AsyncBaseService):
    async def list_audit_events(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request("GET", "/v1/audit-events", query=query)

    async def list_members(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return await self._request("GET", "/v1/members", query=query)

    async def create_member(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/members", body=body)

    async def delete_member(self, user_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/members/{userID}", path_params={"userID": user_id},
        )

    async def bulk_member(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/members/bulk", body=body)

    async def list_roles(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return await self._request("GET", "/v1/roles", query=query)

    async def create_role(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/roles", body=body)

    async def get_role(self, role_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/roles/{roleID}", path_params={"roleID": role_id},
        )

    async def update_role(self, role_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "PATCH", "/v1/roles/{roleID}", path_params={"roleID": role_id}, body=body,
        )

    async def delete_role(self, role_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/roles/{roleID}", path_params={"roleID": role_id},
        )

    async def list_resource_policies(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request("GET", "/v1/resource-policies", query=query)

    async def create_resource_policy(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/resource-policies", body=body)

    async def delete_resource_policy(self, policy_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/resource-policies/{policyID}",
            path_params={"policyID": policy_id},
        )

    async def list_tag_policies(
        self, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request("GET", "/v1/tag-policies", query=query)

    async def create_tag_policy(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/tag-policies", body=body)

    async def delete_tag_policy(self, policy_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/tag-policies/{policyID}", path_params={"policyID": policy_id},
        )

    async def seed_roles(self) -> dict[str, Any]:
        return await self._request("POST", "/v1/seed-roles")

File: packages/ruby-sdk/lib/strait/operations/rbac.rb
# frozen_string_literal: true

module Strait
  module Operations
    # Service for role-based access control: members, roles, resource policies, tag policies, and audit events.
    class RBACService < BaseService
      # List audit events.
      # GET /v1/rbac/audit-events
      def list_audit_events(query: nil)
        _request(:get, "/v1/rbac/audit-events", query: query)
      end

      # List members.
      # GET /v1/rbac/members
      def list_members(query: nil)
        _request(:get, "/v1/rbac/members", query: query)
      end

      # Create a member.
      # POST /v1/rbac/members
      def create_member(body)
        _request(:post, "/v1/rbac/members", body: body)
      end

      # Delete a member.
      # DELETE /v1/rbac/members/{userID}
      def delete_member(user_id)
        _request(:delete, "/v1/rbac/members/{userID}", path_params: { "userID" => user_id })
        nil
      end

      # Bulk member operations.
      # POST /v1/rbac/members/bulk
      def bulk_member(body)
        _request(:post, "/v1/rbac/members/bulk", body: body)
      end

      # List roles.
      # GET /v1/rbac/roles
      def list_roles(query: nil)
        _request(:get, "/v1/rbac/roles", query: query)
      end

      # Create a role.
      # POST /v1/rbac/roles
      def create_role(body)
        _request(:post, "/v1/rbac/roles", body: body)
      end

      # Get a role by ID.
      # GET /v1/rbac/roles/{roleID}
      def get_role(role_id)
        _request(:get, "/v1/rbac/roles/{roleID}", path_params: { "roleID" => role_id })
      end

      # Update a role.
      # PATCH /v1/rbac/roles/{roleID}
      def update_role(role_id, body)
        _request(:patch, "/v1/rbac/roles/{roleID}", path_params: { "roleID" => role_id }, body: body)
      end

      # Delete a role.
      # DELETE /v1/rbac/roles/{roleID}
      def delete_role(role_id)
        _request(:delete, "/v1/rbac/roles/{roleID}", path_params: { "roleID" => role_id })
        nil
      end

      # List resource policies.
      # GET /v1/rbac/resource-policies
      def list_resource_policies(query: nil)
        _request(:get, "/v1/rbac/resource-policies", query: query)
      end

      # Create a resource policy.
      # POST /v1/rbac/resource-policies
      def create_resource_policy(body)
        _request(:post, "/v1/rbac/resource-policies", body: body)
      end

      # Delete a resource policy.
      # DELETE /v1/rbac/resource-policies/{policyID}
      def delete_resource_policy(policy_id)
        _request(:delete, "/v1/rbac/resource-policies/{policyID}", path_params: { "policyID" => policy_id })
        nil
      end

      # List tag policies.
      # GET /v1/rbac/tag-policies
      def list_tag_policies(query: nil)
        _request(:get, "/v1/rbac/tag-policies", query: query)
      end

      # Create a tag policy.
      # POST /v1/rbac/tag-policies
      def create_tag_policy(body)
        _request(:post, "/v1/rbac/tag-policies", body: body)
      end

      # Delete a tag policy.
      # DELETE /v1/rbac/tag-policies/{policyID}
      def delete_tag_policy(policy_id)
        _request(:delete, "/v1/rbac/tag-policies/{policyID}", path_params: { "policyID" => policy_id })
        nil
      end

      # Seed default roles.
      # POST /v1/rbac/seed
      def seed_roles
        _request(:post, "/v1/rbac/seed")
      end
    end
  end
end
